include/exadg/time_integration/time_int_abm_base.h

```diff
@@ -0,0 +1,266 @@
+    // correct operator for next time step
+    evaluate_operator(evaluated_operators_np, solution, get_next_time());
+
+    // In case the mesh is moving, the mesh has to be updated between prediction and correction.
+    // Pushing back the vectors here makes it possible to move the mesh before do_timestep_solve(),
+    // thus this class does not have to know how to move the mesh.
+    if(vec_evaluated_operators.size() > 0)
+    {
+      push_back(vec_evaluated_operators);
+      std::swap(vec_evaluated_operators[0], evaluated_operators_np);
+    }
+
+    // predict next timestep
+    predrict_solution(prediction, solution, vec_evaluated_operators);
+
+    // write output
+    if(print_solver_info() && !is_test)
+    {
+      pcout << std::endl << "Adams-Bashforth-Moulton:";
+      print_wall_time(pcout, timer.wall_time());
+    }
+
+    timer_tree->insert({"Timeloop", "Adams-Bashforth-Moulton"}, timer.wall_time());
+  }
+
+  void
+  evaluate_operator(VectorType & dst, VectorType const & src, double const t) const
+  {
+    pde_operator->evaluate(dst, src, static_cast<Number>(t));
+    pde_operator->apply_inverse_mass(dst, dst);
+  }
+
+  void
+  correct_solution(VectorType &                    dst,
+                   VectorType const &              op_np,
+                   std::vector<VectorType> const & ops) const
+  {
+    dst.add(static_cast<Number>(get_time_step_size() * am.get_gamma0()), op_np);
+    for(unsigned int i = 0; i < this->am.get_order() - 1; ++i)
+      dst.add(static_cast<Number>(get_time_step_size() * am.get_alpha(i)), ops[i]);
+  }
+
+  void
+  predrict_solution(VectorType &                    dst,
+                    VectorType const &              src,
+                    std::vector<VectorType> const & ops) const
+  {
+    dst = src;
+    for(unsigned int i = 0; i < this->ab.get_order(); ++i)
+      dst.add(static_cast<Number>(get_time_step_size() * ab.get_alpha(i)), ops[i]);
+  }
+
+  void
+  prepare_vectors_for_next_timestep() final
+  {
+    // vectors have been pushed back inside the time integraton loop
+  }
+
+  void
+  read_restart_vectors(boost::archive::binary_iarchive & ia) final
+  {
+    ia >> solution;
+    ia >> prediction;
+  }
+
+  void
+  write_restart_vectors(boost::archive::binary_oarchive & oa) const final
+  {
+    oa << solution;
+    oa << prediction;
+  }
+
+  virtual void
+  postprocessing() const = 0;
+
+  void
+  solve_steady_problem() final
+  {
+    AssertThrow(false, dealii::ExcMessage("Steady not implemented."));
+  }
+
+  // spatial pde operator
+  std::shared_ptr<Operator> pde_operator;
+
+  // Time integration constants
+  ABTimeIntegratorConstants ab;
+  AMTimeIntegratorConstants am;
+
+  // solution vector
+  VectorType solution;
+  // temporary vector to store prediction
+  VectorType prediction;
+
+  // store evaluated operators from previous time steps
+  VectorType              evaluated_operators_np;
```

Comment:
Regarding the naming, it should be `evaluated_operator_np`, since it is only one time vector/operator in my opinion.

Reply:
[ebc3e2a](https://github.com/exadg/exadg/pull/585/commits/ebc3e2a9db5d954fc665222adf52a981ac34070c)